scalanet/src/io/iohk/scalanet/peergroup/udp/StaticUDPPeerGroup.scala

```diff
@@ -0,0 +1,444 @@
+package io.iohk.scalanet.peergroup.udp
+
+import cats.effect.concurrent.{Ref, Semaphore}
+import cats.effect.Resource
+import cats.implicits._
+import com.typesafe.scalalogging.StrictLogging
+import io.iohk.scalanet.monix_subject.ConnectableSubject
+import io.iohk.scalanet.peergroup.{Channel, Release}
+import io.iohk.scalanet.peergroup.{InetMultiAddress}
+import io.iohk.scalanet.peergroup.Channel.{ChannelEvent, MessageReceived, DecodingError, UnexpectedError}
+import io.iohk.scalanet.peergroup.PeerGroup.ServerEvent.ChannelCreated
+import io.iohk.scalanet.peergroup.ControlEvent.InitializationError
+import io.iohk.scalanet.peergroup.InetPeerGroupUtils.toTask
+import io.iohk.scalanet.peergroup.PeerGroup.{
+  ServerEvent,
+  TerminalPeerGroup,
+  MessageMTUException,
+  ChannelAlreadyClosedException
+}
+import io.netty.bootstrap.Bootstrap
+import io.netty.buffer.Unpooled
+import io.netty.channel.{
+  RecvByteBufAllocator,
+  ChannelOption,
+  ChannelInitializer,
+  ChannelHandlerContext,
+  ChannelInboundHandlerAdapter
+}
+import io.netty.channel.nio.NioEventLoopGroup
+import io.netty.channel.socket.DatagramPacket
+import io.netty.channel.socket.nio.NioDatagramChannel
+import java.io.IOException
+import java.net.{InetSocketAddress, PortUnreachableException}
+import monix.eval.Task
+import monix.execution.Scheduler
+import scala.util.control.NonFatal
+import scodec.{Codec, Attempt}
+import scodec.bits.BitVector
+
+/**
+  * PeerGroup implementation on top of UDP that uses the same local port
+  * when creating channels to remote addresses as the one it listens on
+  * for incoming messages.
+  *
+  * This makes it compatible with protocols that update the peer's port
+  * to the last one it sent a message from.
+  *
+  * It also means that incoming messages cannot be tied to a specific channel,
+  * so if multiple channels are open to the same remote address,
+  * they will all see the same messages. The incoming responses will also
+  * cause a server channel to be opened, where response type messages have
+  * to be discarded, and the server channel can be discarded if there's no
+  * request type message for a long time.
+  *
+  * @tparam M the message type.
+  */
+class StaticUDPPeerGroup[M] private (
+    config: StaticUDPPeerGroup.Config,
+    workerGroup: NioEventLoopGroup,
+    isShutdownRef: Ref[Task, Boolean],
+    serverSubject: ConnectableSubject[ServerEvent[InetMultiAddress, M]],
+    serverChannelsRef: Ref[Task, Map[InetSocketAddress, StaticUDPPeerGroup.ChannelAlloc[M]]],
```

Comment:
why do we need separate refs for server channels and client channels ? If lets say peer1 opens client channer to peer2 , and peer2 sends message to peer1, shoudn't this message lands on this client channel and create new server channel ?

Reply:
That sounds correct, if peer1 opens a client to peer2 and sends a message, then peer2 will receive it on a new server channel. If peer1 sends more messages they arrive on this channel without a new server channel being created. So to handle incoming requests all you have to do is spawn a handler fiber for each new server channel. The responses peer2 sends on that server channel will go to peer1, and here it's going to be a bit unintuitive: peer1 will see the message on a new server channel as well as the client channel it has open. This is because there's generally no way to say why peer2 is sending a message.

If now peer2 opens a client to peer1, it will have one or more client channels too. They will all have the messages replicated, so it can use that client to just collect the messages that look like responses to what it sent on that client.

We need to keep these separate because for an incoming message there must be at most one server channel to handle all requests, so it can do things like accumulate state. If there's already an outgoing client channel connected to the same address, it may not expect the message coming in right now, so while we have to send it there, it's unlikely to handle it properly. 

Something like this:

```scala
peer1.server.handle {
  case ChannelCreated(channel, release) =>
    channel.in.collect {
       case MessageReceived(m: Request) => 
          channel.sendMessage(response(m))
    }
   .guarantee(release)
   .startAndForget
}

peer1.client(peer2.processAddress).use {
  client.sendMessage(request) >> client.in.collect {
     case MessageReceived(m: Response) if m.requestId == request.requestId =>
       m
  }.headL.timeout(5.seconds)
}
```

Reply:
I see. I am not sure about this but this whole abstraction it seems complex for general purpose udp thing as:
- it is pretty unintuitive
- it also makes some assumptions about how it will be used (request/response)
I would say that that we need next set of eyes on this, and maybe it is time to consider simple wrapper over netty udp like:
```
trait Udp {
 send(m): Tasl[Unit]
 incomining: Observble[M]
}
```
and go from there, as i fear we are just trying to fit square peg in round hole forcoing udp to fit in this channel abstraction.

Reply:
Yes this is the only way to make it work with this abstraction and the kind of protocol Ethereum is using, i.e. there's no request-id in responses. That simple trait you are proposing is how Netty works, however if we used that, then the same kind of channel-replication machinery would have to be done in the application level, for example the the [ExpectedResponseChannels](https://github.com/ethereum/trinity/blob/dc0ddbb5ee5129b29ea2d4a9444f6a1b60cf09f4/p2p/discovery.py#L1352) in `trinity` and the [replyMatcher](https://github.com/ethereum/go-ethereum/blob/master/p2p/discover/v4_udp.go#L94) in `go-ethereum`.

The current approach appeals to me because even though it might be unintuitive, the actual code that is required to handle stuff looks exactly the same as we do for the other implementation, e.g. [KNetwork.sendRequest](https://github.com/input-output-hk/scalanet/blob/develop/scalanet/kademlia/src/io/iohk/scalanet/kademlia/KNetwork.scala#L100). So you don't have to be aware, as long as you have a reasonable protocol where you can distinguish request and responses.

Reply:
It wouldn't be unreasonable to think that at some point we might want to add options to make this class more knowledgeable and able to pair up channels, but the making it more efficient, while keeping the interface. But if we give up on the `PeerGroup` abstraction I'm not sure what reason we have left to bother with Scalanet in this case.

Reply:
On the positive side https://github.com/input-output-hk/scalanet/pull/93 worked with almost no changes, the only fix needed is easy to understand.